### ISSUE TITLE
BUG: `np.array()` assumes that bytestrings are ASCII, if mixed with strings

### ISSUE THREAD

**Issue by Zac-HD** [numpy/numpy#23899]
TL;DR: either `np.array(["str", b"bytes"])` should be an error for mixing string types, or `np.array(["str", b"non-ascii: \xff"])` should not crash.  This is probably low priority though.

```pycon
$ python  # version 3.10
>>> import numpy as np
>>> np.__version__
'1.23.5'

# Looks like we coerce input to string types, whether unicode or bytes
>>> np.array([1.337, "str"])
array(['1.337', 'str'], dtype='<U32')
>>> np.array([1.337, b"bytes"])
array([b'1.337', b'bytes'], dtype='|S32')

# and if we _mix_ unicode and bytes (uh-oh!), we silently get unicode
>>> np.array(["str", b"bytes"])
array(['str', 'bytes'], dtype='<U5')

# but this groundlessly assumes that the bytestring is valid ASCII, and boom:
>>> np.array(["str", b"non-ascii: \xff"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 11: ordinal not in range(128)
```

This is pretty surprising, because in general I expect bytestrings can contain _arbitrary_ bytes, e.g. packed structs, and Numpy is generally pretty good about that.  Plausibly related issues:

- https://github.com/numpy/numpy/issues/12075
- https://github.com/numpy/numpy/issues/6315
- https://github.com/numpy/numpy/issues/4162#issuecomment-32602496
- https://github.com/HypothesisWorks/hypothesis/pull/3670#discussion_r1222564734

**Comment by jobh** [comment]
ASCII aside, I would expect a string/bytes mixture to end up as a bytes array, since all strings can be promoted to bytes but not vice versa (unicode or not)?

**Comment by Zac-HD** [comment]
> ASCII aside, I would expect a string/bytes mixture to end up as a bytes array, since all strings can be promoted to bytes but not vice versa (unicode or not)?

All bytestrings can be interpreted as _latin-1_-encoded text, and all unicode strings can be encoded as utf-32 bytes, but neither of these is compatible with utf-8 and that's probably what users expect.

**Comment by ngoldbaum** [comment]
There’s historical baggage here in that unicode support was added to numpy along with python3 support, and originally the bytes type *was* the string type. Just pointing that out if what numpy is doing seems weird.

Honestly, mixing strings and bytes like that is unusual enough that if we were going to change numpy’s behavior i’d say we should error in that case and not try to guess about encodings for the user.
